### PR TITLE
feat: Webhook for IMC

### PR DESCRIPTION
**Pull request by Arvindthiru**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/webhook/validation/uservalidation.go [hunk 1]
```diff
@@ -77,11 +79,43 @@ func ValidateMemberClusterUpdate(currentMC, oldMC fleetv1alpha1.MemberCluster, w
...
+	namespacedName := types.NamespacedName{Name: currentIMC.Name, Namespace: currentIMC.Namespace}
+	imcSpecUpdated := isInternalMemberClusterSpecUpdated(currentIMC.Spec, oldIMC.Spec)
+	imcStatusUpdated := isInternalMemberClusterStatusUpdated(currentIMC.Status, oldIMC.Status)
+	if !imcSpecUpdated && imcStatusUpdated {
+		var mc fleetv1alpha1.MemberCluster
+		if err := client.Get(ctx, types.NamespacedName{Name: namespacedName.Name}, &mc); err != nil {
+			// fail open, if the webhook cannot get member cluster resources we don't block the request.
+			klog.V(2).ErrorS(err, "failed to get member cluster resource for request to modify internal member cluster, allowing request to be handled by api server",
+				"user", userInfo.Username, "groups", userInfo.Groups, "namespacedName", namespacedName)
+			return admission.Allowed("failed to list member cluster resources for request to modify internal member cluster")
```
Comment (ryanzhang-oss) [inline pkg/webhook/validation/uservalidation.go:99]: this msg could use some refining

File: pkg/webhook/validation/uservalidation.go [hunk 2]
```diff
@@ -97,6 +131,16 @@ func isMemberClusterMapFieldUpdated(currentMCLabels, oldMCLabels map[string]stri
...
 }
 
+// isInternalMemberClusterSpecUpdated returns true if internal member cluster spec is updated.
+func isInternalMemberClusterSpecUpdated(currentIMCSpec, oldIMCSpec fleetv1alpha1.InternalMemberClusterSpec) bool {
+	return !reflect.DeepEqual(currentIMCSpec, oldIMCSpec)
+}
+
+// isInternalMemberClusterStatusUpdated returns true if internal member cluster status is updated.
+func isInternalMemberClusterStatusUpdated(currentIMCStatus, oldIMCStatus fleetv1alpha1.InternalMemberClusterStatus) bool {
+	return !reflect.DeepEqual(currentIMCStatus, oldIMCStatus)
```
[the remark this answers is not in the document]
Comment (Arvindthiru) [reply pkg/webhook/validation/uservalidation.go:126]: Resource version is the same when webhook receives the request

File: pkg/webhook/validation/uservalidation.go [hunk 3]
```diff
@@ -77,11 +79,43 @@ func ValidateMemberClusterUpdate(currentMC, oldMC fleetv1alpha1.MemberCluster, w
...
+		if mc.Spec.Identity.Name == userInfo.Username || fmt.Sprintf(serviceAccountFmt, mc.Spec.Identity.Name) == userInfo.Username {
+			klog.V(2).InfoS(fleetResourceAllowed, "user", userInfo.Username, "groups", userInfo.Groups, "kind", imcKind, "namespacedName", namespacedName)
+			return admission.Allowed(fmt.Sprintf(fleetResourceAllowedFormat, userInfo.Username, userInfo.Groups, "InternalMemberCluster", namespacedName))
+		}
+		klog.V(2).InfoS(IMCStatusUpdateNotAllowed, "user", userInfo.Username, "groups", userInfo.Groups, "kind", imcKind, "namespacedName", namespacedName)
+		return admission.Denied(fmt.Sprintf(IMCStatusUpdateNotAllowedFormat, userInfo.Username, userInfo.Groups, namespacedName))
+	} else if imcSpecUpdated && !imcStatusUpdated {
+		return ValidateUserForFleetResource(currentIMC.Kind, namespacedName, whiteListedUsers, userInfo)
+	} else if imcSpecUpdated && imcStatusUpdated {
+		// no user is allowed to update IMC spec & status at the same time.
```
[the remark this answers is not in the document]
Comment (Arvindthiru) [reply pkg/webhook/validation/uservalidation.go]: It's not allowed on a update https://github.com/kubernetes/kubernetes/blob/6e0cb243d57592c917fe449dde20b0e246bc66be/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L130

File: test/e2e/utils/helper.go [hunk 4]
```diff
@@ -273,10 +274,53 @@ func CreateResourcesForWebHookE2E(ctx context.Context, hubCluster *framework.Clu
 	gomega.Eventually(func() error {
 		return hubCluster.KubeClient.Create(ctx, &rb)
 	}, PollTimeout, PollInterval).Should(gomega.Succeed(), "failed to create role binding %s for webhook E2E", rb.Name)
+
+	// Creating this MC for IMC E2E, this MC will fail to join since it's name is not configured to be recognized by the member agent.
```
Comment (ryanzhang-oss) [inline test/e2e/utils/helper.go:278]: I don't get this comment. Member agent doesn't watch MC.

File: test/e2e/utils/helper.go [hunk 5]
```diff
@@ -273,10 +274,53 @@ func CreateResourcesForWebHookE2E(ctx context.Context, hubCluster *framework.Clu
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-mc",
+		},
+	}
+	gomega.Expect(hubCluster.KubeClient.Delete(ctx, &mc)).Should(gomega.Succeed())
+
+	gomega.Eventually(func() error {
+		var imc fleetv1alpha1.InternalMemberCluster
+		return hubCluster.KubeClient.Get(ctx, types.NamespacedName{Name: "test-mc", Namespace: "fleet-member-test-mc"}, &imc)
+	}, PollTimeout, PollInterval).Should(gomega.HaveOccurred(), "Failed to wait for internal member cluster %s to be deleted in %s cluster", "test-mc", hubCluster.ClusterName)
```
Comment (ryanzhang-oss) [inline test/e2e/utils/helper.go:322]: we are looking for a "NotFound" error. Why not make sure it's the case here?

File: test/e2e/utils/helper.go [hunk 6]
```diff
@@ -273,10 +274,53 @@ func CreateResourcesForWebHookE2E(ctx context.Context, hubCluster *framework.Clu
...
+			Identity:               identity,
+			State:                  fleetv1alpha1.ClusterStateJoin,
+			HeartbeatPeriodSeconds: 60,
+		},
+	}
+	gomega.Eventually(func() error {
+		return hubCluster.KubeClient.Create(ctx, mc)
+	}, PollTimeout, PollInterval).Should(gomega.Succeed(), "Failed to wait for member cluster %s to be created in %s cluster", mc.Name, hubCluster.ClusterName)
+
+	imc := &fleetv1alpha1.InternalMemberCluster{
```
Comment (ryanzhang-oss) [inline test/e2e/utils/helper.go:299]: where is this used ?

Comment (Arvindthiru) [reply test/e2e/utils/helper.go:299]: https://github.com/Azure/fleet/blob/7cae41bbb51c01af76cba390cfe38e483453af98/test/e2e/webhook_test.go#L710 in the E2E tests

File: test/e2e/framework/cluster.go [hunk 7]
```diff
@@ -64,7 +65,7 @@ func GetClusterClient(cluster *Cluster) {
 	gomega.Expect(err).Should(gomega.Succeed(), "Failed to set up Rest Mapper")
 
 	cluster.ImpersonateKubeClient, err = client.New(impersonateRestConfig, client.Options{Scheme: cluster.Scheme})
-	gomega.Expect(err).Should(gomega.Succeed(), "Failed to set up Kube Client")
+	gomega.Expect(err).Should(gomega.Succeed(), "Failed to set up Impersonate Kube Client")
```
[the remark this answers is not in the document]
Comment (Arvindthiru) [reply test/e2e/framework/cluster.go:68]: Currently it's being used to simulate a test user who is not part of the system:masters group to make requests to the api server. AFAIK we need to create new Impersonate clients if we want to test scenarios where user has a different name and belongs to different groups

File: pkg/webhook/validation/uservalidation.go [hunk 8]
```diff
@@ -77,11 +71,36 @@ func ValidateMemberClusterUpdate(currentMC, oldMC fleetv1alpha1.MemberCluster, w
...
+		}
+		// For the upstream E2E we use hub agent service account's token which allows member agent to modify IMC status, hence we use serviceAccountFmt to make the check.
+		if mc.Spec.Identity.Name == userInfo.Username || fmt.Sprintf(serviceAccountFmt, mc.Spec.Identity.Name) == userInfo.Username {
+			klog.V(2).InfoS("user in groups is allowed to modify fleet resource", "user", userInfo.Username, "groups", userInfo.Groups, "kind", imcKind, "namespacedName", namespacedName)
+			return admission.Allowed(fmt.Sprintf(fleetResourceAllowedFormat, userInfo.Username, userInfo.Groups, "InternalMemberCluster", namespacedName))
+		}
+		klog.V(2).InfoS("user is not allowed to update IMC status", "user", userInfo.Username, "groups", userInfo.Groups, "kind", imcKind, "namespacedName", namespacedName)
+		return admission.Denied(fmt.Sprintf(imcStatusUpdateNotAllowedFormat, userInfo.Username, userInfo.Groups, namespacedName))
+	}
+	return ValidateUserForFleetResource(currentIMC.Kind, namespacedName, whiteListedUsers, userInfo)
```
[the remark this answers is not in the document]
Comment (Arvindthiru) [reply pkg/webhook/validation/uservalidation.go:101]: As of now we don't pass the label changes from MC to IMC. Agreed label changes won't affect generation

File: test/e2e/framework/cluster.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply test/e2e/framework/cluster.go:68]: yes, we need that. Can you add a TODO?

File: test/e2e/utils/helper.go (hunk 6 above)
Comment (ryanzhang-oss) [reply test/e2e/utils/helper.go:299]: I am a bit confused, the line you pasted declared a new variable.

**Review by ryanzhang-oss — APPROVED** [review]

**Pull request merged**